### PR TITLE
Manual mask mk2

### PR DESCRIPTION
**Pull request by tacaswell**

### CODE REVIEW

**Comment by tacaswell** [comment]
This is ready for review

**Comment by ericdill** [comment]
Sorry that I'm leaving comments on the commits. I'm using the github app for Android and I can not do in line on the total diff

File: xray_vision/mask/manual_mask.py [hunk 1]
```diff
@@ -56,7 +56,7 @@
 
 class ManualMask(object):
     @ensure_ax_meth
-    def __init__(self, ax, image, cmap='gray'):
+    def __init__(self, ax, image, cmap='gray', max_memory=20):
```
Comment (danielballan) [inline xray_vision/mask/manual_mask.py]: max_memory -> max_history or undo_history

File: xray_vision/mask/manual_mask.py [hunk 2]
```diff
@@ -78,27 +81,35 @@ def __init__(self, ax, image, cmap='gray'):
...
+            Defaults to 20.
 
         Attributes
         ----------
         mask : boolean array
             all "postive" regions are True, negative False
+
         label_array : integer array
             each contiguous region is labeled with an integer
-        label_by_stroke : integer array
```
Comment (danielballan) [inline xray_vision/mask/manual_mask.py]: I liked this. Why kill?

**Comment by danielballan** [comment]
I like shift flipping the sign of the lasso.

File: xray_vision/mask/manual_mask.py (hunk 2 above)
Comment (danielballan) [inline xray_vision/mask/manual_mask.py]: Eh, on second thought it was cute but probably not that useful, and I can see how it's tricky to fit into the new pixel-flipping feature.

Comment (tacaswell) [inline xray_vision/mask/manual_mask.py]: I think a better way to implement this is to be able to maintain and show multiple masks, but that it is a much messier UX problem.

**Comment by tacaswell** [comment]
I had wanted to use the right mouse button to do the inversion, but only paying attention to the left mouse button is hard-coded in do `Lasso` :disappointed:

**Comment by tacaswell** [comment]
but it has been like that since 2006 when JDH added it  c8f36bc5ac74e4504dafc9e636cf3cdc7e10777d.

**Comment by danielballan** [comment]
"alt" is actually the traditional choice in Photoshop / Paint Shop Pro /
etc.

File: xray_vision/mask/manual_mask.py (hunk 2 above)
Comment (sameera2004) [inline xray_vision/mask/manual_mask.py]: Yes showing multiple will be hard, But I think it is good this way

**Comment by sameera2004** [comment]
Nice I like the new changes

**Comment by tacaswell** [comment]
I think I have addressed everything.  Anything else?

**Comment by tacaswell** [comment]
duh, the example I haven't finished yet!

**Comment by ericdill** [comment]
Would it be painful to implement shift as a modifier that draws a straight line from the last point to the current mouse position?

**Comment by tacaswell** [comment]
Do you mean draw that as a line of pixels or to do that as the side of a
polygon?  Adding a polygon mode would not be too hard, probably a few hours
of my time.

I also have code that lets you click off points and it will do a best-fit
spline to those points which would be a couple of hours to integrate.

The end game here is exposing a path with Bezier control points.

There is also all of the selection widgets in skimage which we should look
at using.

Those last three things all require a slightly more sophisticated UI as you
need to be able do modal input.

**Comment by ericdill** [comment]
Yeah,  I guess I was thinking of something along the lines of the lines with bezier control points.   Sometimes you actually want to have polygon edges though (think beam stop arm holder), so being able to drag the corners of a polygon without curvy edges would be useful too

**Comment by danielballan** [comment]
I would love to see that integrated, but I think we should reuse others'
work here as much as possible. That sounds really hard to get right.

**Comment by sameera2004** [comment]
Earlier I tried to do something like Image_inspector (https://github.com/blink1073/image_inspector)- @ericdill  recommended this. They have all the options like lines, ellipse, polygons, rectangles, etc....

 However, @tacaswell  didn't like the idea of having UI for manual mask,better to have  a auto mask, and Andrei didn't like the idea of having auto mask... etc.. Then I decided only have simple one, earlier I tried  a version of polygon too.

 Take a look at Image_inspector

**Comment by tacaswell** [comment]
I was apparently confused, I thought that was much more tightly coupled into skimage and their qt framework.

**Comment by sameera2004** [comment]
Yes, it was not easy and I had issues opening new matplotlib libraries in my mac?. However, they have all sort of options. Yes I think selection widgets in skimage should be more helpful @tacaswell

**Comment by tacaswell** [comment]
I am in favor of merging this as-is to get something functional in front of users and then adding in the bells and whistles they ask for.

@sameera2004 I meant the opposite, the tools in skimage are (to my understanding) very tightly coupled to some qt objects they have that manage inter-widget communication.  the `image_inspector` project is completely decoupled from skimage/qt which is why it is more useful.

**Comment by sameera2004** [comment]
@tacaswell ok sure, Thanks for the clarification, yes I found image inspector more useful that is what I mentioned that we can take  a look at that. Anyway I got confused with your earlier comment, As I told I had issues with opening some -mac issues. It took while to figure them...

Yes,  it is good if we can add more options like, line, polygon, etc..later..
However, it will only take few of your hours to add them !!  

Yes it is good we have this for users,

**Pull request merged**